docs/Agent/user-guide/man_page.md

```diff
@@ -399,33 +459,63 @@ Show this message and exit.
 
 ---
 
-**NAME**  
-`pbench-results-move` - move results directories to the server to a 1.0 Pbench Server
+**NAME**
+
+`pbench-results-move` - move results directories to a 1.0 Pbench Server
+
+**SYNOPSIS**
 
-**SYNOPSIS**  
 `pbench-results-move [OPTIONS]`
 
-**DESCRIPTION**  
-This command uploads one or more result directories to the configured v1.0 Pbench Server. The specified API Key is used to authenticate the user and to establish ownership of the data on the server. Once the upload is complete, the result directories are, by default, removed from the local system.
+**DESCRIPTION**
+
+This command uploads one or more result directories to a 1.0 Pbench Server.
+
+Two modes are supported:
+
+1. The results are pushed directly to a Pbench Server using the API Key
+authentication token specified by `--token`, and will be owned by that user.
+The Pbench Server URI can be specified with `--server`, or will be defaulted
+from the active configuration file.
+2. The results are pushed to a Relay server, which may be anywhere reachable
+both from the Pbench Agent host executing the command and a 1.0 Pbench Server.
+The command will report a URI, which can be presented to a 1.0 Pbench Server
+through the `relay` API or from the Pbench Dashboard to cause the server to
+pull the results from the Relay server.
+
+Once the upload is complete, the result directories
+are, by default, removed from the local system.
```

Comment:
When the second `PUT` completes, it's done. Why? Both styles of push are fully synchronous, so that's not an issue.

And we're not even documenting `pbench-results-push`, which operates on tar files rather than on result trees. `push` doesn't have an option or logic to delete the tar file when done.

Reply:
> we're not even documenting `pbench-results-push`

Sorry, I meant `pbench-results-move`...I just misspelled it.  😉 

> When the second PUT completes, it's done. Why? Both styles of push are fully synchronous, so that's not an issue.

I guess it depends on one's perspective:  from the perspective of the `pbench-results-move` command, arguably its job is done when it reports the URI after successfully uploading the files to the relay; however, from the user's perspective, the job's not done until the results are safely on the Pbench Server.

So, I as a user would not want my result directories to be deleted before they are on the Pbench Server, and so triggering a deletion based on them having been uploaded just to the relay is premature.  That is, if there is no feedback mechanism from the Pbench Server to the Agent, it would be better if the default were `--no-delete` when `--relay` is provided.

(As you know, I've been contemplating a mechanism by which the Pbench Server would cause the results to be deleted from the relay and even cause the relay to shut down when the results have been successfully uploaded; these effects would be visible to the Pbench Agent which would allow it to then delete the local results directories...but, in the absence of that mechanism, I would be concerned that the user's results might become lost in transit if they are prematurely removed by the Agent.  😒)

Reply:
Messing with option defaults based on other options is inherently confusing, hard to document, and messy to implement since it takes logic out of the simple click decorators and puts it into the post-parse processing or callbacks.

And this "these effects would be visible to the Pbench Agent" is concerning, too, as there's no continually active "agent" component outside the actual test run. We'd need to invent a new and complicating command to figure this out later; I don't find that idea very pleasant at all. At best, you'd need to identify a relay and a server, use that unimplemented "relay directory" API and then try to match each to a server dataset, possibly across users.

I'm thinking of adding a `POST /api/v1/relay/<uri>?delete=true` that will trigger a relay `DELETE` operation for the tarball and manifest on completed intake; but figuring out when an agent can remove the results tree? That's messy.

Reply:
> Messing with option defaults based on other options

Agreed (although, I think Click has ways of doing it which aren't atrocious).  However, having a user lose their results because they were deleted locally and then the relay disappeared before they were uploaded to the Pbench Server would be Very Bad™.  At the same time, I think we agree that, for a direct connection to the Pbench Server, having the default be `--delete` is a good and useful thing.  So, that leaves us in a dilemma.  And, I think the least bad approach is to make the deletion default be conditional.

> there's no continually active "agent" component outside the actual test run

If (and, I acknowledge, at this point, it's a big "if") we have the Agent command create the relay server (which is probably what we want in the overwhelming majority of cases where we use a relay), the the command can certainly detect when the relay exits (and, hopefully, we can arrange the exit code to be useful).  Failing that, the Agent command knows what URI(s) it did `PUT`'s to; we should be able to arrange for the Agent to poll for the continued-presence/sudden-absence of those resources (e.g., using `HEAD` requests or some other light HTTP protocol), which would allow it to determine when the upload to the Pbench Server is complete without resorting to the "directory" API and/or a separate Agent command.

> I'm thinking of adding a `POST /api/v1/relay/<uri>?delete=true` that will trigger a relay `DELETE` operation for the tarball and manifest on completed intake; but figuring out when an agent can remove the results tree? That's messy.

I agree that the situation on the Agent side is tricky, because it is isolated from the Server.  I think the answer lies in the relay, since that's the common touch point between the Server and the Agent.

However, we do have the manifest file, which could be used to convey additional instructions to the Server.  For instance, we could add a section which tells the Server how to acknowledge the upload is complete by `PUT`'ing an a resource back on the relay.  Once the Agent can retrieve that, it knows that the upload is complete, and it can delete the local result files, delete the relay files, and shut down the relay server.

I'm disinclined to add explicit API support to the Pbench Server for deleting the items from the relay...that's not really (in my opinion) a "Pbench Server Thing".  Beyond that, the Agent (or user) can do that directly; if we want to, we can provide the means to do that transparently via the URIs which the Agent provides to the Server (i.e., the Agent can control what the Server sends to the relay); and, as I suggested above, we could do something via the manifest file, rather than invoking a separate API.

Reply:
> > Messing with option defaults based on other options
> 
> Agreed (although, I think Click has ways of doing it which aren't atrocious).

It's a callback, which moves the defaulting logic out of the decorator and therefore makes it somewhat harder to follow. I did that for the `--token` requirement out of necessity, but I'm not really crazy about doing more of that. I really don't like conditional default, because they're inevitably going to catch people by surprise. The other option would be to always default to `--no-delete` and let them specify `--delete` if they really want it gone. But that's a pain, too.

> However, having a user lose their results because they were deleted locally and then the relay disappeared before they were uploaded to the Pbench Server would be Very Bad™. At the same time, I think we agree that, for a direct connection to the Pbench Server, having the default be `--delete` is a good and useful thing. So, that leaves us in a dilemma. And, I think the least bad approach is to make the deletion default be conditional.

Since the file relay doesn't have any persistent context other than the designated file directory, I assume that if it were to fail you could just restart with the same settings and all would be well. This is basically no different from saying that if the Agent controller filesystem disappears you'll lose the results trees, and I don't see how that's a problem. (Or, any more of a problem than "basic life" in a distributed world.) We have three subsystems, and each are built with certain persistence assumptions and should operate reliably within the scope of those assumptions: I don't see any need to build extra complicated mechanism to have one backstop another.

> > there's no continually active "agent" component outside the actual test run
> 
> If (and, I acknowledge, at this point, it's a big "if") we have the Agent command create the relay server (which is probably what we want in the overwhelming majority of cases where we use a relay), the the command can certainly detect when the relay exits (and, hopefully, we can arrange the exit code to be useful). Failing that, the Agent command knows what URI(s) it did `PUT`'s to; we should be able to arrange for the Agent to poll for the continued-presence/sudden-absence of those resources (e.g., using `HEAD` requests or some other light HTTP protocol), which would allow it to determine when the upload to the Pbench Server is complete without resorting to the "directory" API and/or a separate Agent command.

Sound bizarrely complicated, and I'm not sure it presents much real value. But there's a lot of routes we can think about through the clouds once we get off the ground with this PR.

> > I'm thinking of adding a `POST /api/v1/relay/<uri>?delete=true` that will trigger a relay `DELETE` operation for the tarball and manifest on completed intake; but figuring out when an agent can remove the results tree? That's messy.
> 
> I agree that the situation on the Agent side is tricky, because it is isolated from the Server. I think the answer lies in the relay, since that's the common touch point between the Server and the Agent.
> 
> However, we do have the manifest file, which could be used to convey additional instructions to the Server. For instance, we could add a section which tells the Server how to acknowledge the upload is complete by `PUT`'ing an a resource back on the relay. Once the Agent can retrieve that, it knows that the upload is complete, and it can delete the local result files, delete the relay files, and shut down the relay server.

Again, sounds way overcomplicated for the value, but it's not unreasonable to put that on the table for [PBENCH-1175](https://issues.redhat.com/browse/PBENCH-1175)

> I'm disinclined to add explicit API support to the Pbench Server for deleting the items from the relay...that's not really (in my opinion) a "Pbench Server Thing". Beyond that, the Agent (or user) can do that directly; if we want to, we can provide the means to do that transparently via the URIs which the Agent provides to the Server (i.e., the Agent can control what the Server sends to the relay); and, as I suggested above, we could do something via the manifest file, rather than invoking a separate API.

We're building Relay integration into the server; so unless you're arguing that we shouldn't do that, I don't really see how you can argue that it shouldn't be able to delete the resource it pulled.

In any case, this is another possible idea for [PBENCH-1175](https://issues.redhat.com/browse/PBENCH-1175) and I'm not suggesting adding that to this PR.

Reply:
> > > Messing with option defaults based on other options
> > 
> > Agreed (although, I think Click has ways of doing it which aren't atrocious).
> 
> It's a callback, which moves the defaulting logic out of the decorator

I concur that using the callback is one approach.  But there are a few others.  For instance, after the parsing is done, I think the code can use [`click.Context.get_parameter_source()`](https://click.palletsprojects.com/en/8.1.x/api/#click.Context.get_parameter_source) to determine whether the value of the `--delete/--no-delete` option was specified explicitly on the command line; if not, then the code can decide to use a value of "no delete" if `--relay` was specified.  (And, that behavior is not that hard to explain in the documentation.)

> Since the file relay doesn't have any persistent context other than the designated file directory, I assume that if it were to fail you could just restart with the same settings and all would be well. 

My assumption is that the file relay, and its local storage, is entirely transient.  I imagine one option being to run it in a cloud allocation, which evaporates entirely once it's shut down (e.g., in an on-demand, "serverless" execution, like AWS Lambda).  So, I'm working from the assumption that the results need to persist on the host with the Agent until they have been safely conveyed all the way to the Pbench Server.

Also, I think it would be beneficial in most cases to hide (or abstract-away) the existence of the relay (to the maximum extent feasible).  For instance, we should limit the coupling between the Pbench Server and the relay to just the URIs which the Agent supplies directly to the user and to the Pbench Server via the manifest file; likewise, we should see if we can arrange it so that the user doesn't actually (have to) interact with the relay on the Agent side, either (i.e., unless specifically requested by the user, we can have the Agent create and shutdown the relay automatically and invisibly to the user).  (And, I would like to avoid precluding the possibility of using a fixed service, like Amazon S3 as the relay.)

> We're building Relay integration into the server; so unless you're arguing that we shouldn't do that, I don't really see how you can argue that it shouldn't be able to delete the resource it pulled.

What I think/hope we're building into the Pbench Server is the ability to "pull" results as well as having them "pushed".  That ability is built on top of some sort of relay support, but I would stop short of saying that we're "building relay integration into the _server_" -- the _integration_ that I'm expecting us to provide will be with the _Agent_.

I'm expecting that, in the typical case, the Agent will stand up the relay when the user wants to push a result, and, therefore, it makes a certain amount of sense that the onus should be on the Agent to tear it down when the transfer is complete.  Yes, there is an alternative model where the Agent exits when the results have been uploaded to the relay, and the relay exits when the results have been uploaded to the Server, but I'm not sure that that model is better than keeping the Agent running until the full transfer is complete.  So, if the relay is run under the auspices of the Agent, then there is no need for the Pbench Server to offer the capability of removing results from it -- the Agent can do that if we want it done (although, having the Server remove them as a signal to the Agent that the transfer is complete remains an interesting idea...but we don't need a Server API to drive that).

Reply:
Exactly where which Agent-entangled component "continues to run" and how it manages to monitor relay transfer progress is an enormous open issue. A lot of cases we know about suggest that the host on which the agent runs won't be accessible to the server any more than the server is accessible to it: they need a neutral external party. So that can be a Pbench Relay container or a Pbench Agent Relay container, or just a Pbench Agent command inside a standard Pbench Agent container, but its management and lifetime are disconnected by definition from both the agent and server workflow. This is roughly analogous to the current Pbench Satellite, albeit potentially more general.

Yeah, transient relays are great if you can run `pbench-results-move` on a system that can be reached from the Pbench Server, but having the command sit there with a "type ^C to terminate" is ugly, as is hiding a relay server in the background where it's easy to overlook.

And all of that is exactly why I tore off these two bounded tasks to integrate basic relay capability into both server and agent. We can worry about defining the rest later.

Right now, and until we've defined and built all that, what happens to the relay files? Right now, they exist until the relay is shut down, which is probably OK only if it really *is* transient. Adding a mechanism for the server to delete when it's finished pulling makes sense in this model, as it's the only place we *do* know when the transfer is done.

What to do with the results directory is another question, of course. Do we make someone go back and clean it out manually after doing the server pull? We don't have any active Agent component to monitor the status, nor is it at all clear to me that's the right model. We have `pbench-clear-results` but no way to remove *specific* datasets other than to find the directory in `/var/lib/pbench-agent` and manually `rmdir`, which sucks. While it's true that theoretically provides a fallback if the relay fails, we don't support it well right now.

I'm not at all sure I like "undefaulting" `--delete` for `--relay`, and the code is inevitably a bit ugly. That, at least, I'll think about. The rest is all vague futures and the discussion really doesn't belong in this PR where it'll be lost in the mists of time soon enough...